### ISSUE TITLE
detaching qz_x params in dreg objective

### ISSUE THREAD

**Issue by vandana-rajan** [iffsid/mmvae#4]
Hi,

In the dreg function, why do you detach the encoder params using the following line of code?

>> qz_x = model.qz_x(*[p.detach() for p in model.qz_x_params])  # stop-grad for \phi

Does it mean that the encoder parameters are not updated during training?

Thanks,
VR

**Comment by iffsid** [comment]
> Does it mean that the encoder parameters are not updated during training?

Not quite. 
That particular line ensures that gradients for `\phi` are not computed through the _score function_ `\log q(z | x)` as per [Doubly Reparametrised Gradient Estimators for Monte Carlo Objectives](https://arxiv.org/pdf/1810.04152.pdf). Gradients for `\phi` through the (reparametrised) samples `zs` are still computed as normal. Following the DReG paper, the `dreg` function (as opposed to `_dreg`) actually does the correction to account for this missing gradient.

See [here](https://hackmd.io/@5pwCvlLhSMm2E1skjPTOTQ/elbo) for a more grounded explanation of what the DReG ELBO does.

**Comment by vandana-rajan** [comment]
Thanks.
